hypothesis-python/tests/cover/test_stateful.py

```diff
@@ -1333,9 +1438,157 @@ def use_flatmap(self, bun):
+            assert isinstance(bun, int)
+            assert bun < -1
+            return -bun
+
+        @rule(
+            target=buns,
+            bun=buns.flatmap(lambda x: just(x + 1)).map(lambda x: -x),
+        )
+        def use_flatmap_mapped(self, bun):
+            assert isinstance(bun, int)
+            assert bun < 0
+            return -bun
+
+        @rule(bun=buns)
+        def use_directly(self, bun):
+            assert isinstance(bun, int)
+            assert bun >= 0
+
+    Machine.TestCase.settings = Settings(stateful_step_count=5, max_examples=10)
+    run_state_machine_as_test(Machine)
+
+
+def test_map_with_combinations():
+    class Machine(RuleBasedStateMachine):
+        buns = Bundle("buns")
+
+        @initialize(target=buns)
+        def create_bun(self):
+            return 1
+
+        @rule(bun=buns.map(lambda x: -x))
+        def use_map_base(self, bun):
+            assert isinstance(bun, int)
+            assert bun < 0
+
+        @rule(
+            bun=buns.map(lambda x: -x).filter(lambda x: x < -1),
+        )
+        def use_flatmap_filtered(self, bun):
+            assert isinstance(bun, int)
+            assert bun < -1
+
+        @rule(
+            bun=buns.map(lambda x: -x).flatmap(lambda x: just(abs(x) + 1)),
+        )
+        def use_flatmap_mapped(self, bun):
+            assert isinstance(bun, int)
+            assert bun > 0
+
+        @rule(bun=buns)
+        def use_directly(self, bun):
+            assert isinstance(bun, int)
+            assert bun > 0
+
+    Machine.TestCase.settings = Settings(stateful_step_count=5, max_examples=10)
+    run_state_machine_as_test(Machine)
+
+
+def test_filter_with_combinations():
+    class Machine(RuleBasedStateMachine):
+        buns = Bundle("buns")
+
+        @initialize(target=buns)
+        def create_bun(self):
+            return multiple(0, -1, -2)
+
+        @rule(bun=buns.filter(lambda x: x > 0))
+        def use_filter_base(self, bun):
+            assert isinstance(bun, int)
+            assert bun > 0
+
+        @rule(
+            bun=buns.filter(lambda x: x > 0).flatmap(lambda x: just(-x)),
+        )
+        def use_filter_flatmapped(self, bun):
+            assert isinstance(bun, int)
+            assert bun < 0
+
+        @rule(
+            bun=buns.filter(lambda x: x < 0).map(lambda x: -x),
+        )
+        def use_flatmap_mapped(self, bun):
+            assert isinstance(bun, int)
+            assert bun > 0
+
         @rule(bun=buns)
         def use_directly(self, bun):
             assert isinstance(bun, int)
 
     Machine.TestCase.settings = Settings(stateful_step_count=5, max_examples=10)
     run_state_machine_as_test(Machine)
+
+
+# TODO: This test might be covered by the other `filter` tests,
+# but might be useful in maintaining stateful healthcheck behavior?
+# TODO: Conjecture data doesn't reset as each rule is run, which
+# leads to "1 successful example", and the rest being failures, which might not be ideal?
```

Comment:
I think this is a misdiagnosis: conjecture _shouldn't_ reset between rules anyway.  What you're seeing is that because the initialize rule doesn't draw anything and the only other rule fails, the only way for the test to pass is if we decide to stop immediately after initialize - so that happens once, then we filter everything else out.